Allow promise to be returned from 'add()'

dchambers opened the pull request:
FWIW, this is a partial fix for https://github.com/storybooks/storybook/issues/713.

There are no docs, I haven't promisified the other test functions in storyshot (only `snapshotWithOptions`), and there are no unit tests.

dchambers committed:
Allow promise to be returned from 'add()'

kanzelm3 commented:
+1 this feature would be useful

ndelangen reviewed (approved):
Code looks good

not tested yet

Please resolve linting errors

dchambers committed:
Merge branch 'master' into master

ndelangen committed:
Merge branch 'master' into master

ndelangen committed:
Merge branch 'master' into master

Hypnosphi committed:
Merge branch 'master' into master

ndelangen commented:
Hey @dchambers code looks good to me, and with a clear use-case described [here](https://github.com/storybooks/storybook/issues/713#issuecomment-323844440) I think we can get this rolling.

I think a unit-test or 2 to test this, a note of this in the docs somewhere would be nice.
An example usage in a kitchen-sink (that is snapshotted) is required to get this merged.

ndelangen committed:
Merge branch 'master' into master

ndelangen reviewed (changes requested):
We need an example usage that is snapshotted in a kitchen sink

JasonCoal commented:
What is this?   
  
  
  

  
  

  
  
>   
> On Aug 22, 2017 at 5:25 AM,  <Norbert de Langen (mailto:notifications@github.com)>  wrote:
>   
>   
>   
>
> @ndelangen  requested changes on this pull request.
>
>   
>
> We need an example usage that is snapshotted in a kitchen sink
>
>   
>
> —
>  You are receiving this because you are subscribed to this thread.
>  Reply to this email directly,   view it on GitHub (https://github.com/storybooks/storybook/pull/1670#pullrequestreview-57716395), or   mute the thread (https://github.com/notifications/unsubscribe-auth/AOXMGHhYTlvLzp0hbUS2RBb0O1sA9qLKks5sap6RgaJpZM4O6MYU).
>
>   
>   
>     
>   
>   
>

dchambers committed:
Merge branch 'master' into master

dchambers committed:
Document ability to wrap stories in promises

dchambers commented:
Hi @ndelangen, I've taken quite a good look at the tests that currently exist and can't find anything that looks like it would be able to verify this new functionality. I guess part of the problem is that what's there at present seems to be more like unit tests rather than BDD tests that test the actual behaviour, and in this case unit testing that some promises were added wouldn't seem to really prove anything.

If I were to add someting then the `app/react/src/client/preview/client_api.test.js` test would seem to be the place for a test of `add()` though I would have thought? Any ideas here?

Regarding docs, I have now added a description of the feature that will hopefully be helpful to people.

Thanks again!

ndelangen commented on docs/pages/basics/writing-stories/index.md:
```diff
@@ -1,7 +1,8 @@
----
+* * *
```
Can you revert these lines?, sorry, it's not your fault, the markdown linter needs a patch, But I haven't found the time to understand what changed/broke.

ndelangen commented:
I think the fastest way to get coverage and confidence is to add a story to a kitchen sink that actually uses this feature!

ndelangen committed:
Merge branch 'release/3.3' into master

ndelangen commented:
@dchambers Are you interested in finishing this PR by processing the review comments I left for you? Anything you need from me? Anything unclear?

Please let me know, I can take over if you'd like.

dchambers commented:
Hi @ndelangen, yes apologies, I've been crazy busy with work the last two weeks, including this weekend. I'm fairly confident I'll be able to make some time to work through the issues you've identified on Monday, if not Sunday.

Thanks, Dominic 👍

dchambers commented:
Apologies again for not progressing this at the start of this week as promised. I've still been crazy busy with work. I've now definitely got time to dedicate to this on Friday and Saturday though, so will get everything resolved then. Thanks for your patience on this one 👍

ndelangen committed:
Merge branch 'release/3.3' into master

ndelangen commented:
Rebased for you 👍 , can you check everything is as it should be?

Hypnosphi committed:
Merge branch 'release/3.3' into master

dchambers committed:
Create CRA app for ES6-Promise kitchen sink

This is a completely vanilla CRA app except that it has an `.eslintrc`
file to avoid linting errors about create-react-app itself.

dchambers committed:
Add Storybook to kitchen-sink

dchambers committed:
Use storyshots to test the stories

dchambers committed:
Wrap stories in promises

dchambers committed:
Update kitchen-sink to test code in this repo

Also fixed a subtle bug in the code that only showed up on running
`npm run build-packs`.

dchambers committed:
Make the `README.md` relevant for the kitchen-sink

dchambers commented:
> Rebased for you 👍 , can you check everything is as it should be?

Thanks @ndelangen. It's not quite working yet, but I've just pushed an attempt at a kitchen-sink that is hopefully in the right direction.

ndelangen committed:
Merge branch 'release/3.3' into master

Hypnosphi commented:
Why a separate kitchen-sink?

dchambers commented:
> Why a separate kitchen-sink?

Good point @Hypnosphi, some storybook tests where one of the stories uses a promise could just as easily be added to the existing `cra-kitchen-sink`. @ndelangen, do you have an opinion here?

One question though, unlike some of the other kitchen-sinks, the CRA one appears to be testing versioned releases of storybook (e.g. `"@storybook/addon-actions": "^3.2.0",` within `package.json`), rather than the repo itself. Is this intentional, or is there some magic here to make it test the current repo?

Hypnosphi commented:
This magic is called lerna: https://lernajs.io
It links multiple packages between each other when you run `yarn bootstrap -- --core`

We prefer to keep al examples in one kitchen-sink

Hypnosphi committed:
Merge branch 'release/3.3' into master

ndelangen committed:
Merge branch 'release/3.3' into master

ndelangen commented:
This is broken in combination with decorators... Looking for a possible solution.

ndelangen commented:
Found a solution, I think

ndelangen committed:
ADD support for async components + decorators && ADD async render to cra-kitchen-sink

IMPROVE the in-code comments

ndelangen committed:
FIX documentation & CORRECT usage of async story

A story really should be a function, otherwise all request would happen
immediately upon visiting storybook, this should only happen when visiting that particular story.

ndelangen committed:
DELETE redundant example

ndelangen reviewed (commented):
So the decorator stuff needs to be reviewed and tested really carefully!!!

There's a few alternatives:
- Render stories twice to detect if it's an async story
- have an api specific for async stories
- do not support decorators for async stories (harder than you think)

What's the risk here?
Well this changes the order of execution of storyFn and decorators.
the storyFn gets the context object earlier then before this change, the decorators get the context object after. I'm unsure what the impact of this really is.
Context has always been -blegh- to me.

What this would like break/change in behavior is if someone was doing this: (which I haven't seen anyone do)
```js
storiesOf('Aliens')
  .add(context => <Alien name={'jimmy'} stuff={context} />)
```

Please advice @Hypnosphi @storybooks/team

ndelangen commented:
Obviously unit-tests needs fixing / storyshots needs a fix.

I hope we can have some more eyes looking for on the potential impact on this!
And then I or @dchambers can fix them.

tmeasday commented:
Is this something we should be considering alongside the mooted changes to the addon API?

Hypnosphi commented:
What was the initial point of passing `story` to decorators as a function? Maybe it's supposed that the decorator can somehow affect that function's output before calling it?

Hypnosphi commented:
Maybe in `4.0.0` we should introduce some redux middleware-like API for decorators:
```
addDecorator(context => next => story => ...)
```

Then async stories could be supported with promise decorator:
```
addDecorator(context => next => story => Promise.resolve(story).then(next))
```

danielduan commented:
Closing this because there hasn't been any activity here for a while. Feel free to reopen when this is ready for discussion again.

breathe commented:
Having trouble following this -- is there still no support for any form of async story definition?

leojh commented:
I got confirmation that this feature was never finished. What is the appetite to get this effort started again? Or is there a plan for this to be inherent in the v4 release?

husayt commented:
I have been trying this in 5.2 and it seems this is still not supported. Seems there is an alternative constructs available for using async data in stories. Maybe someone suggest something?

This is what I tried and it doesn't work as add doesn't support promises
```
storiesOf("Article", module)
  .add("test", async () => {
    let dt= await getDynData("article")
    return {
      components: { Article },
      data() {
        return {
          art: {
            link: "test",
            title: dt,
            desc: "test",
          }
        }
      },
      template: `<Article v-bind="art" />`
    }
  })
```

Hypnosphi commented:
@husayt in `5.2-beta`, you can try using Storybook Hooks. They a similar to [React Hooks](https://reactjs.org/docs/hooks-intro.html) but work with any framework:

```
import {useState, useEffect} from '@storybook/hooks'

storiesOf("Article", module)
  .add("test", () => {
    const [dt, setDt] = useState('Loading...')
    useEffect(() => {
      let mounted = true
      getDynData("article").then(result => {
        if (mounted) {
          setDt(result)
        }
      })
      return () => {
        mounted = false
      }
    })

    return {
      components: { Article },
      data() {
        return {
          art: {
            link: "test",
            title: dt,
            desc: "test",
          }
        }
      },
      template: `<Article v-bind="art" />`
    }
  })
```

You may want to extract a custom hook to reuse this kind of logic:

```
function useAcyncData(getData, placeholder) {
  const [data, setData] = useState(placeholder)
  useEffect(() => {
    let mounted = true
    getData().then(result => {
      if (mounted) {
        setData(result)
      }
    })
    return () => {
      mounted = false
    }
  })
  return data
}

storiesOf("Article", module)
  .add("test", () => {
    const data = useAsyncData(() => getDynData("article"), 'Loading...')

    return {
      components: { Article },
      data() {
        return {
          art: {
            link: "test",
            title: dt,
            desc: "test",
          }
        }
      },
      template: `<Article v-bind="art" />`
    }
  })
```

husayt commented:
@Hypnosphi Thank you for direction. I have been playing with this since then, but seems it is not ready yet.

This is what I ended up with:
```
import { useEffect, useState } from "@storybook/client-api"
const dt, setDt
storiesOf("Article", module)
  .add("test", () => {
  console.log("rendering started", dlist)
   ;[dt, setDt] = useState('Loading...')
    useEffect(() => {
      getDynData("article").then(result => {
          setDt(result)
      })
    },[])

    return {
      components: { Article },
      data() {
        return {
          art: {
            link: "test",
            title: dt,
            desc: "test",
          }
        }
      },
      template: `<Article v-bind="art" />`
    }
  })
```
Let me go through issues I faced.

1. Seems this has been updated and I should use `client-api` for import
`import { useEffect, useState } from "@storybook/client-api"`
2. Above example was causing endless loop as `setData` would cause rerender and on each one dt would be undefined. So I had to 
 a. define `dt`, `setDt` outside story, so that value would be retained. 
 b . Also, instead of `mounted` logic which you put to tackle infinite calling, I am passing `[]` as second param to `useEffect` and that ensures its called only once.

But still after all this data doesn't get updated on the Vue component. Vue is not getting updates to `dt`. Seems, there needs to be some kind of bridge to pass reactive data from `dt` to Vue. 

I see  ` console.log("rendering started", dlist)` is being called twice, first time with `..Loading`  and second time with the value from api, but component is only rendered once thus only  mounted once. Updates to `dt` don't make any difference to the component.

So seems like these are Vue related issues. Most probably we are jumping ahead of the line with that, as nowhere it is mentioned that this is expected to work with Vue.

BUt I can see how extremely powerful this can be

Hypnosphi commented:
1. Sure, sorry
2. Just `b.` should be enough. You still need that `mounted` logic to avoid errors when you switch story before the data arrives
